bcgov/biohubbc
# SIMSBIOHUB-470: Observation Table Measurement Validation

## Links to Jira Tickets

[SIMSBIOHUB-470](https://apps.nrs.gov.bc.ca/int/jira/browse/SIMSBIOHUB-470)

## Description of Changes

- Added front end validation for measurement columns in the observations table
- Refactored validation process and moved files into own validation utils file
- Added measurement validation at API entry level for observation measurements as well
- Fixed bugs in import/ measurement validation process

## Testing Notes
### Setup
- Run Critterbase locally from branch `SIMSBIOHUB-498`
- In SIMS, navigate to the `Manage Observations` table on a survey 
- Add records for Moose (Alces Americanus) and Black capped Chickadee (Poecile atricapillus)
- Save those records 
- Open table Cog menu and add measurements
- Search for antler and add all measurements found
- Search for nest and add all measurements found

### Testing
- Add antler values to the chickadee rows 
- Save changes 
- See validation errors for each cell that has antler data

- Add nest values to Moose rows
- Save Changes
- See validation errors for each cell that has nest data

- Add nest data to the chickadee rows
- Add antler data to the moose rows
- Save changes
- See table refresh with saved data
